contrib/coredns/policy/policy.go

```diff
@@ -272,139 +270,170 @@ func (p *PolicyPlugin) decodeDebugMsg(r *dns.Msg) (*dns.Msg, bool) {
 	case typeRedirect:
-		return p.redirect(ctx, w, r, ah.redirect.Value)
+		sendExtra = true
+		status, err = p.redirect(ctx, r, ah.redirect.Value)
 	case typeBlock:
-		return p.retRcode(w, r, dns.RcodeNameError, nil)
+		sendExtra = true
+		status = dns.RcodeNameError
 	case typeRefuse:
-		return p.retRcode(w, r, dns.RcodeRefused, nil)
+		status = dns.RcodeRefused
+	default:
+		status = dns.RcodeServerFailure
+		err = errInvalidAction
 	}
 
-	return p.retRcode(w, r, dns.RcodeServerFailure, errInvalidAction)
+Exit:
+	r.Rcode = status
+	r.Response = true
+	if p.TapIO != nil {
+		if pw := dnstap.NewProxyWriter(w); pw != nil {
+			pw.WriteMsg(r)
+			var attr []*pb.Attribute
+			if sendExtra {
+				attr = ah.attributes()
+			}
+			p.TapIO.SendCRExtraMsg(pw, attr)
+
+		}
+	} else {
+		w.WriteMsg(r)
+	}
+	return status, err
 }
 
 // Name implements the Handler interface
 func (p *PolicyPlugin) Name() string { return "policy" }
 
-func (p *PolicyPlugin) redirect(ctx context.Context, w dns.ResponseWriter, r *dns.Msg, dst string) (int, error) {
-	state := request.Request{W: w, Req: r}
+func (p *PolicyPlugin) redirect(ctx context.Context, r *dns.Msg, dst string) (int, error) {
 	var rr dns.RR
 	cname := false
 
-	if ipv4 := net.ParseIP(dst).To4(); ipv4 != nil {
+	ip := net.ParseIP(dst)
+	qname, qclass := getNameClass(r)
+	if ipv4 := ip.To4(); ipv4 != nil {
 		rr = new(dns.A)
-		rr.(*dns.A).Hdr = dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeA, Class: state.QClass()}
+		rr.(*dns.A).Hdr = dns.RR_Header{Name: qname, Rrtype: dns.TypeA, Class: qclass}
 		rr.(*dns.A).A = ipv4
-	} else if ipv6 := net.ParseIP(dst).To16(); ipv6 != nil {
+	} else if ipv6 := ip.To16(); ipv6 != nil {
 		rr = new(dns.AAAA)
-		rr.(*dns.AAAA).Hdr = dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeAAAA, Class: state.QClass()}
+		rr.(*dns.AAAA).Hdr = dns.RR_Header{Name: qname, Rrtype: dns.TypeAAAA, Class: qclass}
 		rr.(*dns.AAAA).AAAA = ipv6
 	} else {
 		dst = strings.TrimSuffix(dst, ".") + "."
 		rr = new(dns.CNAME)
-		rr.(*dns.CNAME).Hdr = dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeCNAME, Class: state.QClass()}
+		rr.(*dns.CNAME).Hdr = dns.RR_Header{Name: qname, Rrtype: dns.TypeCNAME, Class: qclass}
 		rr.(*dns.CNAME).Target = dst
 		cname = true
 	}
 
-	a := new(dns.Msg)
-	a.Compress = true
-	a.Authoritative = true
-	a.Answer = []dns.RR{rr}
-
 	if cname {
-		msg := new(dns.Msg)
-		msg.SetQuestion(dst, state.QType())
-
-		responseWriter := nonwriter.New(w)
-		status, err := plugin.NextOrFailure(p.Name(), p.Next, ctx, responseWriter, msg)
+		origName := r.Question[0].Name
+		r.Question[0].Name = dst
+		responseWriter := new(Writer)
+		status, err := plugin.NextOrFailure(p.Name(), p.Next, ctx, responseWriter, r)
+		r.Question[0].Name = origName
 		if err != nil {
-			return p.retRcode(w, r, dns.RcodeServerFailure, err)
+			return dns.RcodeServerFailure, err
 		}
-
-		a.Answer = append(a.Answer, responseWriter.Msg.Answer...)
-		a.SetRcode(r, status)
+		r.Answer = []dns.RR{rr}
```

Comment:
common code for "if" and "else", can be moved before "if" statement

Reply:
r.answer affects on NextOrFailure() so we cannot move this line up